lib/AST/Stmt.cpp

```diff
@@ -440,6 +440,71 @@ void LabeledConditionalStmt::setCond(StmtCondition e) {
   Cond = e;
 }
 
+/// Whether or not this conditional stmt rebinds self with a `let self`
+/// or `let self = self` condition. If `requireLoadExpr` is `true`,
+/// additionally requires that the RHS of the self condition is a `LoadExpr`.
+bool LabeledConditionalStmt::rebindsSelf(ASTContext &Ctx,
+                                         bool requireLoadExpr) const {
+  return llvm::any_of(getCond(), [&Ctx, requireLoadExpr](const auto &cond) {
+    return cond.rebindsSelf(Ctx, requireLoadExpr);
+  });
+}
+
+/// Whether or not this conditional stmt rebinds self with a `let self`
+/// or `let self = self` condition. If `requireLoadExpr` is `true`,
+/// additionally requires that the RHS of the self condition is a `LoadExpr`.
+bool StmtConditionElement::rebindsSelf(ASTContext &Ctx,
+                                       bool requireLoadExpr) const {
+  auto pattern = getPatternOrNull();
+  if (!pattern) {
+    return false;
+  }
+
+  // Check whether or not this pattern defines a new `self` decl
+  bool isSelfRebinding = false;
+  if (pattern->getBoundName() == Ctx.Id_self) {
+    isSelfRebinding = true;
+  }
+
+  else if (auto OSP = dyn_cast<OptionalSomePattern>(pattern)) {
+    if (auto subPattern = OSP->getSubPattern()) {
+      isSelfRebinding = subPattern->getBoundName() == Ctx.Id_self;
+    }
+  }
+
+  if (!isSelfRebinding) {
+    return false;
+  }
+
+  // Check that the RHS expr is exactly `self` and not something else
+  Expr *exprToCheckForDRE = getInitializerOrNull();
+  if (!exprToCheckForDRE) {
+    return false;
+  }
+
+  // The RHS could be a `LoadExpr` or `DeclRefExpr`
+  Expr *loadExprSubExpr = nullptr;
+  if (auto LE = dyn_cast<LoadExpr>(exprToCheckForDRE)) {
```

Comment:
I don't understand these conditions, should this be:

```C++
if (requireLoadExpr && !isa<LoadExpr>(exprToCheckForDRE))
  return false;

if (auto *load = dyn_cast<LoadExpr>(exprToCheckForDRE))
  exprToCheckForDRE = load->getSubExpr();

if (auto *DRE = dyn_cast<DeclRefExpr>(exprToCheckForDRE->getSemanticsProvidingExpr())) {
  auto *decl = DRE->getDecl();
  return decl && decl->hasName() ? decl->getName().isSimpleName(Ctx.Id_self) : false;
}
...
```

Reply:
That's much better, thanks! Updated: https://github.com/apple/swift/pull/65310/commits/c029dd693a4b3ba76251ac8ec41142f5587fe349